core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java

```diff
@@ -106,7 +106,8 @@
 
     static final ChannelFutureListener CLOSE_ON_FAILURE = future -> {
         final Throwable cause = future.cause();
-        if (cause != null && !(cause instanceof ClosedStreamException)) {
+        if (cause != null && (!(cause instanceof ClosedStreamException) ||
+                              cause instanceof ClosedSessionException)) {
```

Comment:
`cause instanceof ClosedSessionException` is `true` when the cause is `ClosedStreamException`.

Reply:
So I think checking if it's subclass or not is better and readable not to make mistakes.
```java
if (cause instanceof ClosedSessionException) {
    return;
}
...
```

Reply:
`if (cause instanceof ClosedSessionException) {` will return `ClosedSessionException` but,
The `ClosedSessionException` have to invoke `safeClose` so we cannot just return `ClosedSessionException`.

The channel will be closed if:
* the cause is not null
* accept all causes except `ClosedStreamException`

How about inverting the boolean expression?
```java
if (cause == null) {
    return;
}
if (cause instanceof ClosedStreamException && !(cause instanceof ClosedSessionException)) {
    return;
}

// or just one if 
if (cause == null || (cause instanceof ClosedStreamException &&
                      !(cause instanceof ClosedSessionException))) {
    return;
}
```

Reply:
Ah, I didn't mean to return when the `cause` is an instance of `ClosedSessionException`, but call `safeClose`.
The original condition was wrong and I think these kinna expression will confuse others as well.

Reply:
So this could be:
```java
static final ChannelFutureListener CLOSE_ON_FAILURE = future -> {
    final Throwable cause = future.cause();
    if (cause == null) {
        return;
    }
    if (cause instanceof ClosedSessionException) {
        logAndSafeClose(future, cause);
        return;
    }
    if (cause instanceof ClosedStreamException) {
        return;
    }
    logAndSafeClose(future, cause);
};

private static void logAndSafeClose(ChannelFuture future, Throwable cause) {
    final Channel ch = future.channel();
    logException(ch, cause);
    safeClose(ch);
}

```

Reply:
And we might change this listener to send RST_STREAM when `cause` is an instance of `ClosedStreamException`.

Reply:
Also, I'm wondering if we can add this listener a little bit more inside such as encoder so that we can send RST_STREAM or close the channel completely.

Reply:
Let me address this comment once we decide our direction. 😀